src/uri/params/serializable.cr

```diff
@@ -0,0 +1,57 @@
+require "uri"
+
+require "./from_form_data"
+
+module URI::Params::Serializable
+  macro included
+    def self.from_form_data(params : ::URI::Params)
+      new_from_form_data(params)
+    end
+
+    # :nodoc:
+    #
+    # This is needed so that nested types can pass the name thru internally.
+    # Has to be public so the generated code can call it, but should be considered an implementation detail.
+    def self.from_form_data(params : ::URI::Params, name : String)
+      new_from_form_data(params, name)
+    end
+
+    protected def self.new_from_form_data(params : ::URI::Params, name : String? = nil)
+      instance = allocate
+      instance.initialize(__uri_params: params, name: name)
+      GC.add_finalizer(instance) if instance.responds_to?(:finalize)
+      instance
+    end
+
+    macro inherited
+      def self.from_form_data(params : ::URI::Params)
+        new_from_form_data(params)
+      end
+
+      # :nodoc:
+      def self.from_form_data(params : ::URI::Params, name : String)
+        new_from_form_data(params, name)
+      end
+    end
+  end
+
+  # :nodoc:
+  def initialize(*, __uri_params params : ::URI::Params, name : String?)
+    {% begin %}
+      {% for ivar, idx in @type.instance_vars %}
+        %name{idx} = name.nil? ? {{ivar.name.stringify}} : "#{name}[#{{{ivar.name.stringify}}}]"
```

Comment:
A top level call to `.from_form_data` will have `name` be `nil`, otherwise we'll append on the additional names to handle nested data. E.g. `filter[direction]`.

Would be possible to change this, but this is what I went with because I tested `['foo' => 'bar','biz' => ['age' => 10,'alive' => true]]` with PHP and it was encoded as `foo=bar&biz[age]=10&biz[alive]=1`, which Crystal parses as `URI::Params{"foo" => ["bar"], "biz[age]" => ["10"], "biz[alive]" => ["1"]}`. 🤷.